File: src/principle/liskov_substitution.ts
class Rectangle {
    private _width: number
    get width() {
        return this._width
    }

    private _height: number
    get height() {
        return this._height
    }

    setWidth(width: number) {
        this._width = width
    }

    setHeight(height: number) {
        this._height = height
    }

    constructor(w: number, h: number) {
        this._width = w
        this._height = h
    }

    print() {
        console.log(`width: ${this._width}, height: ${this._height}`)
    }
}

class Square extends Rectangle {
    setWidth(width: number) {
        super.setWidth(width)
        super.setHeight(width)
    }

    setHeight(height: number) {
        super.setWidth(height)
        super.setHeight(height)
    }

    constructor(size: number) {
        super(size, size);
    }

    print() {
        super.print();
    }
}

const resize = (shape: Rectangle) => {
    shape.print()

    while (shape.width <= shape.height) {
        shape.setWidth(shape.width + 1)
    }

    shape.print()
}

// test
(() => {

    let rect = new Rectangle(6, 8);
    let square = new Square(6)

    // 正确
    resize(rect)
    // width: 6, height: 8
    // width: 8, height: 8

    // 死循环
    resize(square)
})()

File: src/principle/liskov_substitution.rs
trait Shape {
    fn get_width(&self) -> u32;
    fn set_width(&mut self, width: u32);
    fn get_height(&self) -> u32;
    fn set_height(&mut self, height: u32);
    fn print(&self);
}

struct Rectangle {
    width: u32,
    height: u32,
}

impl Rectangle {
    pub fn new(width: u32, height: u32) -> Rectangle {
        Rectangle { width, height }
    }
}

impl Shape for Rectangle {
    fn get_width(&self) -> u32 {
        self.width
    }

    fn set_width(&mut self, width: u32) {
        self.width = width;
    }

    fn get_height(&self) -> u32 {
        self.height
    }

    fn set_height(&mut self, height: u32) {
        self.height = height;
    }

    fn print(&self) {
        println!("width: {}, height: {}", self.get_width(), self.get_height());
    }
}


#[test]
fn liskov_substitution_example() {
    let mut rect = Rectangle::new(6, 8);

    rect.print();
    // width: 6, height: 8

    while rect.get_width() < rect.get_height() {
        rect.set_width(rect.get_width() + 1);
    }

    rect.print();
    // width: 8, height: 8
}